### PR TITLE
[SPARK-10471] [CORE] [MESOS] prevent getting offers for unmet constraints

### PR DESCRIPTION
**Pull request by felixb**
this change rejects offers for slaves with unmet constraints for 120s to mitigate offer starvation.
this prevents mesos to send us these offers again and again.
in return, we get more offers for slaves which might meet our constraints.
and it enables mesos to send the rejected offers to other frameworks.

### CODE REVIEW

**Comment by dragos** [comment]
You should probably modify the fine-grained scheduler in the same way.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 1]
```diff
@@ -56,6 +56,7 @@ private[spark] class CoarseMesosSchedulerBackend(
   with MesosSchedulerUtils {
 
   val MAX_SLAVE_FAILURES = 2     // Blacklist a slave after this many failures
+  val CONSTRAINT_MISMATCH_REJECT_OFFER_DURATION = 120000   // reject offers with mismatched constraints
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Can we make this configurable? Also please comment on the unit as well.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 2]
```diff
@@ -240,48 +241,56 @@ private[spark] class CoarseMesosSchedulerBackend(
...
+              .setName("Task " + taskId)
+              .addAllResources(cpuResourcesToUse.asJava)
+              .addAllResources(memResourcesToUse.asJava)
+
+            sc.conf.getOption("spark.mesos.executor.docker.image").foreach { image =>
+              MesosSchedulerBackendUtil
+                .setupContainerBuilderDockerInfo(image, sc.conf, taskBuilder.getContainerBuilder())
+            }
+
+            // accept the offer and launch the task
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Nit: Capitalize the first word as you did in all other comments.

**Comment by tnachen** [comment]
I think the change makes sense, we're planning to add dynamic attribute changes on the slave but that's not merged yet in Mesos. as @dragos mentioned please add this to coarse grain mode too.
And once you make this a configuration please also update the spark on mesos docs in docs folder .

**Comment by felixb** [comment]
I made the duration configurable. Still need to add it to fine grained scheduler.

**Comment by andrewor14** [comment]
ok to test

File: docs/running-on-mesos.md [hunk 3]
```diff
@@ -194,6 +194,10 @@ conf.set("spark.mesos.constraints", "tachyon=true;us-east-1=false")
 
 For example, Let's say `spark.mesos.constraints` is set to `tachyon=true;us-east-1=false`, then the resource offers will be checked to see if they meet both these constraints and only then will be accepted to start new executors.
 
+Spark rejects offers for for slaves with unmet constraints for 120 seconds by default.
```
Comment (tnachen) [inline docs/running-on-mesos.md]: There is also a configurations.md that you should add this too.

**Comment by felixb** [comment]
Added the same logic for fine grained scheduler.
All mesos configuration is only available in `running-on-mesos.md`. So we skipped adding it to `configurations.md`

**Comment by felixb** [comment]
I just rebased to the current upstream/master.
Any hint, why the tests keep failing?

**Comment by dragos** [comment]
They're probably just flaky.

File: docs/running-on-mesos.md (hunk 3 above)
Comment (dragos) [inline docs/running-on-mesos.md]: @tnachen, none of the Yarn or Mesos specific settings are listed in there.

Comment (dragos) [inline docs/running-on-mesos.md]: Also, typo: `for for`.

**Comment by felixb** [comment]
fixed typo.

**Comment by SleepyThread** [comment]
@tnachen @andrewor14 friendly reminder..

**Comment by tnachen** [comment]
There is a big HTML table in the bottom of this file, can you also add it
to that list?

**Comment by felixb** [comment]
added to table of parameters.

**Comment by SleepyThread** [comment]
@tnachen @andrewor14 friendly reminder..

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 4]
```diff
@@ -244,48 +248,56 @@ private[spark] class CoarseMesosSchedulerBackend(
...
-            .setSlaveId(offer.getSlaveId)
-            .setCommand(createCommand(offer, cpusToUse + extraCoresPerSlave, taskId))
-            .setName("Task " + taskId)
-            .addAllResources(cpuResourcesToUse.asJava)
-            .addAllResources(memResourcesToUse.asJava)
-
-          sc.conf.getOption("spark.mesos.executor.docker.image").foreach { image =>
-            MesosSchedulerBackendUtil
-              .setupContainerBuilderDockerInfo(image, sc.conf, taskBuilder.getContainerBuilder())
+        if (meetsConstraints) {
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: It's getting a bit messy now, but I think we can refactor this later as I'd like to introduce more fine grained reasons around rejecting offers.

**Comment by tnachen** [comment]
Thanks overall this LGTM. I know constraints is not yet supported for the cluster scheduler, so ideally when we add that we should also apply this too.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala (hunk 4 above)
Comment (SleepyThread) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: I had already done some refactoring around this specific area in patch #8771

**Comment by SleepyThread** [comment]
@andrewor14 Can you take a look at this and merge this patch?

**Comment by felixb** [comment]
Is there anything I can do to get this merged?

**Comment by dragos** [comment]
@rxin @andrewor14 can we get your attention please?

**Comment by felixb** [comment]
friendly reminder.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 5]
```diff
@@ -101,6 +101,10 @@ private[spark] class CoarseMesosSchedulerBackend(
   private val slaveOfferConstraints =
     parseConstraintString(sc.conf.get("spark.mesos.constraints", ""))
 
+  // reject offers with mismatched constraints in seconds
+  private val rejectOfferDurationForUnmetConstraints =
+    sc.conf.getDouble("spark.mesos.rejectOfferDurationForUnmetConstraints", 120)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: please use `conf.getTimeAsSeconds` instead, and maybe I would call this `spark.mesos.unmetConstraintsRejectOfferDuration`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 6]
```diff
@@ -244,48 +248,56 @@ private[spark] class CoarseMesosSchedulerBackend(
...
-            Collections.singleton(offer.getId),
-            Collections.singleton(taskBuilder.build()), filters)
         } else {
-          // Decline the offer
-          logDebug(s"Declining offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
-          d.declineOffer(offer.getId)
+          // This offer does not meet constraints. We don't need to see it again.
+          // Decline the offer for a long period of time.
+          logDebug(s"Declining offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus"
+              + s" for $rejectOfferDurationForUnmetConstraints")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: this needs to say `for X seconds`. Right now it looks like `Declining offer for 120`.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 7]
```diff
@@ -63,6 +63,10 @@ private[spark] class MesosSchedulerBackend(
   private[this] val slaveOfferConstraints =
     parseConstraintString(sc.conf.get("spark.mesos.constraints", ""))
 
+  // reject offers with mismatched constraints in seconds
+  private val rejectOfferDurationForUnmetConstraints =
+    sc.conf.getDouble("spark.mesos.rejectOfferDurationForUnmetConstraints", 120)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: can you put this in `trait MesosSchedulerUtils` to avoid duplicating this?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 8]
```diff
@@ -207,23 +211,44 @@ private[spark] class MesosSchedulerBackend(
...
+      val (offersMatchingConstraints, offersNotMatchingConstraints) =
+        offers.asScala.partition { o =>
+          val offerAttributes = toAttributeMap(o.getAttributesList)
+          val meetsConstraints =
+            matchesAttributeRequirements(slaveOfferConstraints, offerAttributes)
+
+          // add some debug messaging
+          if (!meetsConstraints) {
+            val id = o.getId.getValue
+            logDebug(s"Declining offer: $id with attributes: $offerAttributes")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: nit: just inline it
```
logDebug(s"Declining offer: ${o.getId.getValue} with attributes: $offerAttributes")
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 9]
```diff
@@ -207,23 +211,44 @@ private[spark] class MesosSchedulerBackend(
...
   override def resourceOffers(d: SchedulerDriver, offers: JList[Offer]) {
     inClassLoader() {
+      // Fail first on offers with unmet constraints
+      val (offersMatchingConstraints, offersNotMatchingConstraints) =
+        offers.asScala.partition { o =>
+          val offerAttributes = toAttributeMap(o.getAttributesList)
+          val meetsConstraints =
+            matchesAttributeRequirements(slaveOfferConstraints, offerAttributes)
+
+          // add some debug messaging
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: remove (doesn't add any value)

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 10]
```diff
@@ -207,23 +211,44 @@ private[spark] class MesosSchedulerBackend(
...
+        }
+
+      // These offers do not meet constraints. We don't need to see them again.
+      // Decline the offer for a long period of time.
+      offersNotMatchingConstraints.foreach { o =>
+        d.declineOffer(o.getId, Filters.newBuilder()
+          .setRefuseSeconds(rejectOfferDurationForUnmetConstraints).build())
+      }
+
       // Fail-fast on offers we know will be rejected
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: this comment is outdated:
```
// Of the matching constraints, see which ones give us enough memory and cores
```

File: docs/running-on-mesos.md [hunk 11]
```diff
@@ -194,6 +194,10 @@ conf.set("spark.mesos.constraints", "tachyon=true;us-east-1=false")
 
 For example, Let's say `spark.mesos.constraints` is set to `tachyon=true;us-east-1=false`, then the resource offers will be checked to see if they meet both these constraints and only then will be accepted to start new executors.
 
+Spark rejects offers for slaves with unmet constraints for 120 seconds by default.
+This makes mesos not sending the rejected offers again and again.
+The reject offer duration is configurable using `conf.set("spark.mesos.rejectOfferDurationForUnmetConstraints", "60")` (for example).
+
```
Comment (andrewor14) [inline docs/running-on-mesos.md]: I actually don't think this is worth documenting. I can't imagine anyone who wants to set this explicitly as long as we pick a sane enough default.

**Comment by andrewor14** [comment]
@felixb sorry for slipping. This looks pretty good. My comments are mostly minor. Thanks for taking the time to fix this.

**Comment by felixb** [comment]
I worked in all your comments.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala [hunk 12]
```diff
@@ -336,4 +336,8 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
     }
   }
 
+  def getRejectOfferDurationForUnmetConstraints(sc: SparkContext): Long = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala]: please keep this protected

File: docs/running-on-mesos.md [hunk 13]
```diff
@@ -194,6 +194,10 @@ conf.set("spark.mesos.constraints", "tachyon:true;us-east-1:false")
 
 For example, Let's say `spark.mesos.constraints` is set to `tachyon:true;us-east-1:false`, then the resource offers will be checked to see if they meet both these constraints and only then will be accepted to start new executors.
 
+Spark rejects offers for slaves with unmet constraints for 120 seconds by default.
+This makes mesos not sending the rejected offers again and again.
+The reject offer duration is configurable using `conf.set("spark.mesos.rejectOfferDurationForUnmetConstraints", "60")` (for example).
```
Comment (andrewor14) [inline docs/running-on-mesos.md]: it doesn't make sense to document how to set this without documenting the configuration myself. I would just revert the changes here altogether since this seems too low level for the user to care about.

**Comment by felixb** [comment]
Is there anything else I can do?

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
LGTM merging into master and 1.6. Thanks for your work and patience!
